File: project files/server/src/services/emailService.js
const nodemailer = require('nodemailer');
const logger = require('../config/logger');

let transporter = null;

const getTransporter = () => {
  if (transporter) return transporter;
  const host = process.env.SMTP_HOST;
  const port = Number(process.env.SMTP_PORT || 587);
  const user = process.env.SMTP_USER;
  const pass = process.env.SMTP_PASS;

  if (!host || !user || !pass) {
    return null;
  }

  transporter = nodemailer.createTransport({
    host,
    port,
    secure: port === 465,
    auth: { user, pass }
  });
  return transporter;
};

const sendEmail = async ({ to, subject, text, html }) => {
  const tx = getTransporter();
  if (!tx) {
    logger.warn(`SMTP not configured. Email skipped for ${to}: ${subject}`);
    return { queued: false };
  }
  await tx.sendMail({
    from: process.env.SMTP_FROM || process.env.SMTP_USER,
    to,
    subject,
    text,
    html
  });
  return { queued: true };
};

module.exports = { sendEmail };

File: spring_back/src/main/java/com/groceryapp/service/EmailService.java
package com.groceryapp.service;

import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.util.Map;

@Slf4j
@Service
public class EmailService {
    public Map<String, Object> sendEmail(String to, String subject, String text, String html) {
        String host = System.getenv("SMTP_HOST");
        String user = System.getenv("SMTP_USER");
        String pass = System.getenv("SMTP_PASS");
        if (host == null || host.isBlank() || user == null || user.isBlank() || pass == null || pass.isBlank()) {
            log.warn("SMTP not configured. Email skipped for {}: {}", to, subject);
            return Map.of("queued", false);
        }
        // External SMTP transport can be plugged in here.
        log.info("Email queued to {} with subject {}", to, subject);
        return Map.of("queued", true);
    }
}
